d2/src/main/java/com/linkedin/d2/balancer/simple/ClusterLoadBalancerSubscriber.java

```diff
@@ -43,13 +46,28 @@ public ClusterLoadBalancerSubscriber(SimpleLoadBalancerState simpleLoadBalancerS
   @Override
   protected void handlePut(final String listenTo, final ClusterProperties discoveryProperties)
   {
-    // TODO: pick stable or canary configs if canary exists
     if (discoveryProperties != null)
     {
+      ClusterProperties pickedProperties = discoveryProperties;
+      CanaryDistributionProvider.Distribution distribution = CanaryDistributionProvider.Distribution.STABLE;
+      if (discoveryProperties instanceof ClusterStoreProperties) // this should always be true since the serializer returns the composite class
+      {
+        ClusterStoreProperties clusterStoreProperties = (ClusterStoreProperties) discoveryProperties;
+        CanaryDistributionProvider canaryDistributionProvider = _simpleLoadBalancerState.getCanaryDistributionProvider();
+        if (clusterStoreProperties.hasCanary() && canaryDistributionProvider != null)
+        {
+          // Canary config and canary distribution provider exist, distribute to use either stable config or canary config.
+          distribution = canaryDistributionProvider
+              .distribute(CanaryDistributionStrategyConverter.toConfig(clusterStoreProperties.getCanaryDistributionStrategy()));
+        }
+        pickedProperties = clusterStoreProperties.getDistributedClusterProperties(distribution);
+      }
+      // TODO: set canary/stable config metric
+
```

Comment:
nit: Maybe extract this to a separate helper function?

Reply:
sure, cleaner then. done.